File: webapp/src/types/playbook_run.ts
// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {TimelineEvent, TimelineEventType} from 'src/types/rhs';
import {Checklist, isChecklist} from 'src/types/playbook';

export interface PlaybookRun {
    id: string;
    name: string;
    summary: string;
    owner_user_id: string;
    reporter_user_id: string;
    team_id: string;
    channel_id: string;
    create_at: number;
    end_at: number;
    post_id: string;
    playbook_id: string;
    checklists: Checklist[];
    status_posts: StatusPost[];
    current_status: PlaybookRunStatus;
    last_status_update_at: number;
    reminder_post_id: string;
    reminder_message_template: string;
    reminder_timer_default_seconds: number;

    /** Previous reminder timer as nanoseconds */
    previous_reminder: number;
    broadcast_channel_ids: string[];
    timeline_events: TimelineEvent[];
    retrospective: string;
    retrospective_published_at: number;
    retrospective_was_canceled: boolean;
    retrospective_reminder_interval_seconds: number;
    participant_ids: string[];
}

export interface StatusPost {
    id: string;
    create_at: number;
    delete_at: number;
}

export interface Metadata {
    channel_name: string;
    channel_display_name: string;
    team_name: string;
    num_participants: number;
    total_posts: number;
    followers: string[];
}

export interface FetchPlaybookRunsReturn {
    total_count: number;
    page_count: number;
    has_more: boolean;
    items: PlaybookRun[];
}

export enum PlaybookRunStatus {
    InProgress = 'InProgress',
    Finished = 'Finished',
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function isPlaybookRun(arg: any): arg is PlaybookRun {
    return Boolean(arg &&
        arg.id && typeof arg.id === 'string' &&
        arg.name && typeof arg.name === 'string' &&
        typeof arg.description === 'string' &&
        arg.owner_user_id && typeof arg.owner_user_id === 'string' &&
        arg.reporter_user_id && typeof arg.reporter_user_id === 'string' &&
        arg.team_id && typeof arg.team_id === 'string' &&
        arg.channel_id && typeof arg.channel_id === 'string' &&
        typeof arg.create_at === 'number' &&
        typeof arg.end_at === 'number' &&
        typeof arg.post_id === 'string' &&
        arg.playbook_id && typeof arg.playbook_id === 'string' &&
        arg.checklists && Array.isArray(arg.checklists) && arg.checklists.every(isChecklist) &&
        arg.status_posts && Array.isArray(arg.status_posts) && arg.status_posts.every(isStatusPost) &&
        typeof arg.reminder_post_id === 'string' &&
        typeof arg.reminder_message_template === 'string' &&
        typeof arg.reminder_timer_default_seconds === 'number' &&
        arg.broadcast_channel_ids && Array.isArray(arg.broadcast_channel_ids) && arg.broadcast_channel_ids.every(isString) &&
        arg.timeline_events && Array.isArray(arg.timeline_events) && arg.timeline_events.every(isTimelineEvent) &&
        arg.participant_ids && Array.isArray(arg.participant_ids) && arg.participant_ids.every(isString)) &&
        typeof arg.last_status_update_at === 'number' &&
        typeof arg.previous_reminder === 'number';
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function isStatusPost(arg: any): arg is StatusPost {
    return Boolean(arg &&
        arg.id && typeof arg.id === 'string' &&
        typeof arg.create_at === 'number' &&
        typeof arg.delete_at === 'number');
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function isMetadata(arg: any): arg is Metadata {
    return Boolean(arg &&
        arg.channel_name && typeof arg.channel_name === 'string' &&
        arg.channel_display_name && typeof arg.channel_display_name === 'string' &&
        arg.team_name && typeof arg.team_name === 'string' &&
        typeof arg.num_participants === 'number' &&
        typeof arg.total_posts === 'number' &&
        arg.followers && Array.isArray(arg.followers) && arg.followers.every(isString));
}

export function isTimelineEvent(arg: any): arg is TimelineEvent {
    return Boolean(arg &&
        typeof arg.id === 'string' &&
        typeof arg.playbook_run_id === 'string' &&
        typeof arg.create_at === 'number' &&
        typeof arg.delete_at === 'number' &&
        typeof arg.event_at === 'number' &&
        typeof arg.event_type === 'string' && Object.values(TimelineEventType).includes(arg.event_type) &&
        typeof arg.summary === 'string' &&
        typeof arg.details === 'string' &&
        typeof arg.post_id === 'string' &&
        typeof arg.subject_user_id === 'string' &&
        typeof arg.creator_user_id === 'string');
}

function isString(arg: any): arg is string {
    return Boolean(typeof arg === 'string');
}

export function playbookRunIsActive(playbookRun: PlaybookRun): boolean {
    return playbookRun.current_status === PlaybookRunStatus.InProgress;
}

export interface FetchPlaybookRunsParams {
    page: number;
    per_page: number;
    team_id?: string;
    sort?: string;
    direction?: string;
    statuses?: string[];
    owner_user_id?: string;
    participant_id?: string;
    search_term?: string;
    playbook_id?: string;
    active_gte?: number;
    active_lt?: number;
    started_gte?: number;
    started_lt?: number;
}

export interface FetchPlaybookRunsParamsTime {
    active_gte?: number;
    active_lt?: number;
    started_gte?: number;
    started_lt?: number;
}

export const DefaultFetchPlaybookRunsParamsTime: FetchPlaybookRunsParamsTime = {
    active_gte: 0,
    active_lt: 0,
    started_gte: 0,
    started_lt: 0,
};

export const fetchParamsTimeEqual = (a: FetchPlaybookRunsParams, b: FetchPlaybookRunsParamsTime) => {
    return Boolean(a.active_gte === b.active_gte &&
        a.active_lt === b.active_lt &&
        a.started_gte === b.started_gte &&
        a.started_lt === b.started_lt);
};

File: client/playbook_run.go
package client

import "time"

// PlaybookRun represents a playbook run.
type PlaybookRun struct {
	ID                            string          `json:"id"`
	Name                          string          `json:"name"`
	Description                   string          `json:"description"`
	OwnerUserID                   string          `json:"owner_user_id"`
	ReporterUserID                string          `json:"reporter_user_id"`
	TeamID                        string          `json:"team_id"`
	ChannelID                     string          `json:"channel_id"`
	CreateAt                      int64           `json:"create_at"`
	EndAt                         int64           `json:"end_at"`
	DeleteAt                      int64           `json:"delete_at"`
	ActiveStage                   int             `json:"active_stage"`
	ActiveStageTitle              string          `json:"active_stage_title"`
	PostID                        string          `json:"post_id"`
	PlaybookID                    string          `json:"playbook_id"`
	Checklists                    []Checklist     `json:"checklists"`
	StatusPosts                   []StatusPost    `json:"status_posts"`
	ReminderPostID                string          `json:"reminder_post_id"`
	PreviousReminder              time.Duration   `json:"previous_reminder"`
	BroadcastChannelID            string          `json:"broadcast_channel_id"`
	ReminderMessageTemplate       string          `json:"reminder_message_template"`
	InvitedUserIDs                []string        `json:"invited_user_ids"`
	InvitedGroupIDs               []string        `json:"invited_group_ids"`
	TimelineEvents                []TimelineEvent `json:"timeline_events"`
	ExportChannelOnArchiveEnabled bool            `json:"export_channel_on_archive_enabled"`
}

// StatusPost is information added to the playbook run when selecting from the db and sent to the
// client; it is not saved to the db.
type StatusPost struct {
	ID       string `json:"id"`
	CreateAt int64  `json:"create_at"`
	DeleteAt int64  `json:"delete_at"`
}

// PlaybookRunMetadata tracks ancillary metadata about a playbook run.
type PlaybookRunMetadata struct {
	ChannelName        string `json:"channel_name"`
	ChannelDisplayName string `json:"channel_display_name"`
	TeamName           string `json:"team_name"`
	NumMembers         int64  `json:"num_members"`
	TotalPosts         int64  `json:"total_posts"`
}

// TimelineEventType describes a type of timeline event.
type TimelineEventType string

const (
	PlaybookRunCreated TimelineEventType = "incident_created"
	TaskStateModified  TimelineEventType = "task_state_modified"
	StatusUpdated      TimelineEventType = "status_updated"
	OwnerChanged       TimelineEventType = "owner_changed"
	AssigneeChanged    TimelineEventType = "assignee_changed"
	RanSlashCommand    TimelineEventType = "ran_slash_command"
)

// TimelineEvent represents an event recorded to a playbook run's timeline.
type TimelineEvent struct {
	ID            string            `json:"id"`
	PlaybookRunID string            `json:"playbook_run"`
	CreateAt      int64             `json:"create_at"`
	DeleteAt      int64             `json:"delete_at"`
	EventAt       int64             `json:"event_at"`
	EventType     TimelineEventType `json:"event_type"`
	Summary       string            `json:"summary"`
	Details       string            `json:"details"`
	PostID        string            `json:"post_id"`
	SubjectUserID string            `json:"subject_user_id"`
	CreatorUserID string            `json:"creator_user_id"`
}

// PlaybookRunCreateOptions specifies the parameters for PlaybookRunService.Create method.
type PlaybookRunCreateOptions struct {
	Name        string `json:"name"`
	OwnerUserID string `json:"owner_user_id"`
	TeamID      string `json:"team_id"`
	Description string `json:"description"`
	PostID      string `json:"post_id"`
	PlaybookID  string `json:"playbook_id"`
}

// Sort enumerates the available fields we can sort on.
type Sort string

const (
	// SortByCreateAt sorts by the "create_at" field. It is the default.
	SortByCreateAt Sort = "create_at"

	// SortByID sorts by the "id" field.
	SortByID Sort = "id"

	// SortByName sorts by the "name" field.
	SortByName Sort = "name"

	// SortByOwnerUserID sorts by the "owner_user_id" field.
	SortByOwnerUserID Sort = "owner_user_id"

	// SortByTeamID sorts by the "team_id" field.
	SortByTeamID Sort = "team_id"

	// SortByEndAt sorts by the "end_at" field.
	SortByEndAt Sort = "end_at"

	// SortBySteps sorts playbooks by the number of steps in the playbook.
	SortBySteps Sort = "steps"

	// SortByStages sorts playbooks by the number of stages in the playbook.
	SortByStages Sort = "stages"

	// SortByTitle sorts by the "title" field.
	SortByTitle Sort = "title"

	// SortByRuns sorts by the number of times a playbook has been run.
	SortByRuns Sort = "runs"
)

// SortDirection determines whether results are sorted ascending or descending.
type SortDirection string

const (
	// Desc sorts the results in descending order.
	SortDesc SortDirection = "desc"

	// Asc sorts the results in ascending order.
	SortAsc SortDirection = "asc"
)

// PlaybookRunListOptions specifies the optional parameters to the
// PlaybookRunService.List method.
type PlaybookRunListOptions struct {
	// TeamID filters playbook runs to those in the given team.
	TeamID string `url:"team_id,omitempty"`

	Sort      Sort          `url:"sort,omitempty"`
	Direction SortDirection `url:"direction,omitempty"`

	// Status filters by All, Ongoing, or Ended; defaults to All.
	Status Status `url:"status,omitempty"`

	// OwnerID filters by owner's Mattermost user ID. Defaults to blank (no filter).
	OwnerID string `url:"owner_user_id,omitempty"`

	// MemberID filters playbook runs that have this member. Defaults to blank (no filter).
	MemberID string `url:"member_id,omitempty"`

	// SearchTerm returns results of the search term and respecting the other header filter options.
	// The search term acts as a filter and respects the Sort and Direction fields (i.e., results are
	// not returned in relevance order).
	SearchTerm string `url:"search_term,omitempty"`

	// PlaybookID filters playbook runs that are derived from this playbook id.
	// Defaults to blank (no filter).
	PlaybookID string `url:"playbook_id,omitempty"`

	// ActiveGTE filters playbook runs that were active after (or equal) to the unix time given (in millis).
	// A value of 0 means the filter is ignored (which is the default).
	ActiveGTE int64 `url:"active_gte,omitempty"`

	// ActiveLT filters playbook runs that were active before the unix time given (in millis).
	// A value of 0 means the filter is ignored (which is the default).
	ActiveLT int64 `url:"active_lt,omitempty"`

	// StartedGTE filters playbook runs that were started after (or equal) to the unix time given (in millis).
	// A value of 0 means the filter is ignored (which is the default).
	StartedGTE int64 `url:"started_gte,omitempty"`

	// StartedLT filters playbook runs that were started before the unix time given (in millis).
	// A value of 0 means the filter is ignored (which is the default).
	StartedLT int64 `url:"started_lt,omitempty"`
}

// PlaybookRunList contains the paginated result.
type PlaybookRunList struct {
	TotalCount int  `json:"total_count"`
	PageCount  int  `json:"page_count"`
	HasMore    bool `json:"has_more"`
	Items      []*PlaybookRun
}

// Status is the type used to specify the activity status of the playbook run.
type Status string

const (
	StatusReported Status = "Reported"
	StatusActive   Status = "Active"
	StatusResolved Status = "Resolved"
	StatusArchived Status = "Archived"
)

type GetPlaybookRunsResults struct {
	TotalCount int           `json:"total_count"`
	PageCount  int           `json:"page_count"`
	HasMore    bool          `json:"has_more"`
	Items      []PlaybookRun `json:"items"`
	Disabled   bool          `json:"disabled"`
}

// StatusUpdateOptions are the fields required to update a playbook run's status
type StatusUpdateOptions struct {
	Status            Status `json:"status"`
	Description       string `json:"description"`
	Message           string `json:"message"`
	ReminderInSeconds int64  `json:"reminder"`
}
